adal/src/main/java/com/microsoft/aad/adal/Oauth2.java

```diff
@@ -767,7 +834,38 @@ private void stopHttpEvent(final HttpEvent httpEvent) {
                 EventStrings.HTTP_EVENT);
     }
 
-    private void setTokenEndpoint(final String tokenEndpoint) {
-        mTokenEndpoint = tokenEndpoint;
+    private void setTokenEndpoint(final String authority) {
+        try {
+            mTokenEndpoint = getUpdatePreferredNetworkLocation(authority) + DEFAULT_TOKEN_ENDPOINT;
+        } catch(AuthenticationException ex) {
+            mTokenEndpoint = authority + DEFAULT_TOKEN_ENDPOINT;
+        }
+    }
+
+    private String getUpdatePreferredNetworkLocation(final String authority) throws AuthenticationException {
```

Comment:
See `AcquireTokenRequest.java` @ `L259` -- I suggest refactoring for reuse where possible